pkg/odo/cli/component/push.go

```diff
@@ -175,6 +176,28 @@ func (po *PushOptions) Validate() (err error) {
 	// If Devfile is present we do not need to validate the below S2I checks
 	// TODO: Perhaps one day move Devfile validation to here instead?
 	if util.CheckPathExists(po.DevfilePath) {
+		spinner := log.Spinner("Validating devfile component")
+		defer spinner.End(false)
+		err = util.ValidateK8sResourceName("component name", po.EnvSpecificInfo.GetName())
+		if err != nil {
+			return err
+		}
+		containerComponents := adaptersCommon.GetDevfileContainerComponents(po.Devfile.Data)
```

Comment:
we can extract https://github.com/openshift/odo/pull/3959/files#diff-6e3a2602ffd394e9a73d4fbf1388be8dR185-R190 out into a helper function and use it in create as well.

Reply:
```
	containerComponents := adaptersCommon.GetDevfileContainerComponents(devObj.Data)
	for _, comp := range containerComponents {
		err := util.ValidateK8sResourceName("container name", comp.Name)
		if err != nil {
			return err
		}
	}
```

Reply:
updated to use unit tests instead of integration tests